source/common/http/conn_manager_utility.cc

```diff
@@ -92,8 +92,30 @@ ConnectionManagerUtility::MutateRequestHeadersResult ConnectionManagerUtility::m
   if (!Utility::isUpgrade(request_headers)) {
     request_headers.removeConnection();
     request_headers.removeUpgrade();
+
     if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.sanitize_te")) {
-      request_headers.removeTE();
+      auto teHeader = request_headers.getTEValue();
+
+      if (!teHeader.empty()) {
+        auto hasTrailersTE = false;
+
+        auto teValues = absl::StrSplit(, ",");
+        for (const auto& teValue : teValues) {
+          auto parts = absl::StrSplit(teValue, ";"); // Handles cases like "chunked, trailers;q=0.5"
+          auto value = absl::StripAsciiWhitespace(parts[0]);
+
+          if (value == Http::Headers::get().TEValues.Trailers) {
+            hasTrailersTE = true;
+            break;
+          }
+        }
+
+        if (hasTrailersTE) {
+          request_headers.setTE(Http::Headers::get().TEValues.Trailers);
+        } else {
+          request_headers.removeTE();
+        }
+      }
```

Comment:
There's a decent bit of logic in here. Can we extract this it a function named something like `sanitizeTEHeader()` and write unit tests for it?

(The integration tests are GREAT but unit tests are nice to because they're very efficient and very focused)